src/storage-client/src/controller/collection_mgmt.rs

```diff
@@ -78,27 +88,101 @@ impl CollectionManager {
+                                for notif in notifs {
+                                    // We don't care if the listener disappeared.
+                                    let _ = notif.send(Err(StorageError::IdentifierMissing(id)));
+                                }
+                            }
+
+                            // As updates succeed we'll remove them from the set.
+                            while !updates.is_empty() {
+                                // Gather all of the updates into a request for persist.
+                                let request = updates
+                                    .iter()
+                                    .map(|(id, req)| {
+                                        let rows = req
+                                            .rows
+                                            .clone()
+                                            .into_iter()
+                                            .map(|(row, diff)| TimestamplessUpdate { row, diff })
+                                            .collect::<Vec<_>>();
+                                        (*id, rows, T::from(now()))
+                                    })
+                                    .collect();
 
-                            let updates = vec![(id, updates.into_iter().map(|(row, diff)| TimestamplessUpdate {
-                                row,
-                                diff,
-                            }).collect::<Vec<_>>(), T::from(now()))];
+                                // Append updates to persist!
+                                let append_result = write_handle.monotonic_append(request).await.expect("sender hung up");
 
-                            loop {
-                                let append_result = write_handle.monotonic_append(updates.clone()).await.expect("sender hung up");
                                 match append_result {
-                                    Ok(()) => break,
-                                    Err(StorageError::InvalidUppers(ids)) => {
+                                    // Everything was successful!
+                                    Ok(()) => {
+                                        // Notify all of our listeners.
+                                        for (_id, UpdateRequest { notifs, .. }) in updates {
+                                            for notif in notifs {
+                                                // We don't care if the listener went away.
+                                                let _ = notif.send(Ok(()));
+                                            }
+                                        }
+                                        // Break because there are no more updates to send.
+                                        break
+                                    },
+                                    // Failed to write to some collections.
+                                    Err(StorageError::InvalidUppers(failed_ids)) => {
```

Comment:
Just OOC, did you observe this message during your testing?

Reply:
I don't remember exactly, but I don't believe I did